### PR TITLE
Never type unsafe lint improvements

### PR DESCRIPTION
**Pull request by WaffleLapkin**
- Move linting code to a separate method
- Remove mentions of `core::convert::absurd` (#124311 was rejected)
- Make the lint into FCW

The last thing is a bit weird though. On one hand it should be `EditionSemanticsChange(2024)`, but on the other hand it shouldn't, because we also plan to break it on all editions some time later. _Also_, it's weird that we don't have `FutureReleaseSemanticsChangeReportInDeps`, IMO "this might cause UB in a future release" is important enough to be reported in deps...

IMO we ought to have three enums instead of [`FutureIncompatibilityReason`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint_defs/enum.FutureIncompatibilityReason.html#):

```rust
enum IncompatibilityWhen {
     FutureRelease,
     Edition(Edition),
}

enum IncompatibilyWhat {
    Error,
    SemanticChange,
}

enum IncompatibilityReportInDeps {
    No,
    Yes,
}
```

Tracking:
- https://github.com/rust-lang/rust/issues/123748

### CODE REVIEW

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 1]
```diff
@@ -4321,7 +4320,12 @@ declare_lint! {
     /// [`()`]: https://doc.rust-lang.org/core/primitive.unit.html
     pub NEVER_TYPE_FALLBACK_FLOWING_INTO_UNSAFE,
     Warn,
-    "never type fallback affecting unsafe function calls"
+    "never type fallback affecting unsafe function calls",
+    @future_incompatible = FutureIncompatibleInfo {
+        reason: FutureIncompatibilityReason::FutureReleaseSemanticsChange,
+        reference: "issue #123748 <https://github.com/rust-lang/rust/issues/123748>",
+    };
+    report_in_external_macro
```
Comment (compiler-errors) [inline compiler/rustc_lint_defs/src/builtin.rs:4328]: didnt know about this :0

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
Do we already have consensus from T-lang that this is OK to bump to a FCW? I can't remember the policy around that. If you can find something, could you link it?

**Comment by WaffleLapkin** [comment]
I'm not aware of any policies with FCWs, ig we could nominate for T-lang?

**Comment by compiler-errors** [comment]
Yeah, let's do that. Also would like their input if they prefer this start out as an edition warning or an FCW.

**Comment by Nilstrieb** [comment]
> Also, it's weird that we don't have FutureReleaseSemanticsChangeReportInDeps, 

probably intuitively simply because no one has needed it yet

**Comment by traviscross** [comment]
The plan that was FCPed (twice) by the lang team in #123508 is:

1. Prior to the stabilization of Rust 2024:
    - Change fallback to `!` on the nightly 2024 edition.
    - Add a lint against fallback affecting a generic that is passed to an `unsafe` function.
        - Perhaps make this lint `deny-by-default` or a hard error in Rust 2024.
    - Add a future incompat lint for some/most of the code that will be broken by fallback to `!`.
    - Add documentation/migration tools for people adopting the 2024 edition.
2. Stabilize Rust 2024.
3. At a later date when 2024 edition adoption is widespread:
    - Make the breaking change to fall back to `!` always everywhere.
    - Change `Infallible = !`.
    - Stabilize `!` (!).

My read is that the plan implies that this lint (about fallback affecting a generic that is passed to an `unsafe` function) should go immediately to a FCW.  While the second bullet under item 1 does not specify this directly, it's covered by the third bullet, since this lint is also catching code that will be broken by fallback to `!`.  Further, since the plan is to change fallback to `!` always everywhere, that implies that all of these lints should be FCWs.

Put differently, it would not make sense to lint harder against #125289 than against #123939 when the latter is probably the more critical lint.

The meeting minutes from 2024-04-10 also confirm that people wanted these lints to fire in all editions.

**Review by compiler-errors — APPROVED** [review]
OK in that case I think this is good to go

**Pull request merged**